### ISSUE TITLE
Unable to create association when child is an association too and is required

### ISSUE THREAD

**Issue by weyert** [balderdashy/waterline#1235]
I am having a model named Conversation where I have the field named `participants` which is an association with the model `ConversationParticipant` now when I make the `conversation`-field in this model required. The creation of association records is failed. If I removed the `required`-flag from the field it's working without problems.

I am doing something wrong?

If I try to do the following:
```js
Conversation.update({id: 1}, {participants: [
  {conversation: 1, rsvpStatus: "active"
]})
```

I am getting the error:
```
Invalid attributes sent to ConversationParticipant:
 • conversation
   • "required" validation rule failed for input: null
```

Conversation-model: 
```json5
{  schema: true, attributes: {
    participants: {
      collection: 'ConversationParticipant',
      via: 'conversation'
    }
}}
```

ConversationParticipant-model
```json5
{  schema: true,   autoPK: true,
   attributes: {

   conversation: {
      model: 'Conversation'
      required: true
    },

    rsvpStatus: {
      type: 'string',
      required: true,
      defaultsTo: 'inactive',
      in: ['active', 'inactive'']
    },

}}
```

**Comment by weyert** [comment]
Anyone experience this before?

**Comment by devinivy** [comment]
I hope you don't mind, I added code blocks into your post so that I could read it more easily!

This looks like it could be related to the naive strategy used for nested updates ("nested" referring here to updating a collection of participants which live nested on a conversation).  The way nested updates works is that all associations are cleared, then the desired associations are created.  In that first step required foreign keys can be set to `null` and cause an error like the one you're experiencing, and typically the way around that is to not use nested updates.

What confuses me is that `activity` should be the field getting set to `null` (since it's the `via` of the relationship).  I don't see `activity` in the ConversationParticipant model at all though.  Did `activity` get renamed to `conversation` or something?  I think there's a some missing information.

**Comment by weyert** [comment]
Yes, sorry, I have been missing up similar data structures. Yes, `activity` had to be `conversation`. So it's best to not mark this field as required? Or do you think this is a bug in Waterline?

Related question, do you know whether these nested updates are done before beforeUpdate() is being called on the Conversation-model? I am trying to find the right spot to add my code for updating some counters :)

**Comment by devinivy** [comment]
Nested updates occur between `beforeUpdate()` and `afterUpdate()`.

It's hard to say whether this should technically be a bug or not.  Though it's definitely a pain-point that other people have run into.  I guess it's just good to know that nested updates will set existing associated foreign keys to `null` then create associations only for the records listed in the update call.  @particlebanana might have some ideas how this should be handled by waterline.

I suggest keeping the `required` field, and updating associated records directly on the ConversationParticipant collection without using nested updates.

**Comment by sailsbot** [comment]
Thanks for posting, @weyert.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
